File: frontend/src/features/feedback/types/feedback.ts
/**
 * Type definitions for AI-generated interview feedback.
 */

/**
 * Complete feedback data returned from the backend API.
 */
export interface InterviewFeedback {
  id: string;
  session_id: string;

  // Dimension scores (0-100)
  technical_accuracy_score: number;
  communication_clarity_score: number;
  problem_solving_score: number;
  relevance_score: number;
  overall_score: number;

  // Detailed feedback text for each dimension
  technical_feedback: string;
  communication_feedback: string;
  problem_solving_feedback: string;
  relevance_feedback: string;
  overall_comments: string | null;

  // Structured recommendations
  knowledge_gaps: string[];
  learning_recommendations: string[];

  // Metadata
  created_at: string;
  updated_at: string;
}

/**
 * Props for the DimensionScores component.
 */
export interface DimensionScoresProps {
  overallScore: number;
  technicalAccuracy: number;
  communicationClarity: number;
  problemSolving: number;
  relevance: number;
}

/**
 * Props for the KnowledgeGaps component.
 */
export interface KnowledgeGapsProps {
  knowledgeGaps: string[];
}

/**
 * Props for the LearningRecommendations component.
 */
export interface LearningRecommendationsProps {
  learningRecommendations: string[];
}

File: backend/app/schemas/feedback.py
"""
Pydantic schemas for interview feedback API.
"""

import datetime as dt
from typing import List
from pydantic import BaseModel, ConfigDict, UUID4, field_validator


class FeedbackAnalysisResult(BaseModel):
    """Schema for OpenAI feedback analysis JSON response."""

    technical_accuracy_score: int
    communication_clarity_score: int
    problem_solving_score: int
    relevance_score: int
    technical_feedback: str
    communication_feedback: str
    problem_solving_feedback: str
    relevance_feedback: str
    overall_comments: str | None = None
    knowledge_gaps: List[str]
    learning_recommendations: List[str]

    @field_validator(
        "technical_accuracy_score",
        "communication_clarity_score",
        "problem_solving_score",
        "relevance_score",
    )
    @classmethod
    def clamp_score(cls, v: int) -> int:
        """Clamp scores to 0-100 range."""
        return max(0, min(100, v))


class InterviewFeedbackResponse(BaseModel):
    """Response schema for interview feedback."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID4
    session_id: UUID4
    technical_accuracy_score: int
    communication_clarity_score: int
    problem_solving_score: int
    relevance_score: int
    overall_score: int
    technical_feedback: str
    communication_feedback: str
    problem_solving_feedback: str
    relevance_feedback: str
    overall_comments: str | None
    knowledge_gaps: List[str]
    learning_recommendations: List[str]
    created_at: dt.datetime
    updated_at: dt.datetime
